### PR TITLE
Bug 722007 - update row in place in AndroidBrowserHistoryDataExtender.

### PR DESCRIPTION
**Pull request by ncalexan**
And factor out AndroidBrowserHistoryDataExtenderTest.

The ticket asks not to blow away the existing visit data, but I don't know how to do that without changing the history extension database schema: since visits are stored as a JSON string, there's no 'update only part'.  I doubt we want to change the database schema right now, so I'm pushing this small cleaning/preparation for new data extenders patch and suggesting we close/de-prioritize 722007.

Or I've completely misunderstood :)

### CODE REVIEW

**Comment by emtwo** [comment]
Here's the original commit with a few comments: https://github.com/mozilla-services/android-sync/commit/b22ef

File: src/main/java/org/mozilla/gecko/sync/repositories/android/AndroidBrowserHistoryDataExtender.java [hunk 1]
```diff
@@ -46,12 +47,11 @@
 import android.database.Cursor;
 import android.database.sqlite.SQLiteDatabase;
 import android.database.sqlite.SQLiteOpenHelper;
-import android.util.Log;
 
 public class AndroidBrowserHistoryDataExtender extends SQLiteOpenHelper {
 
-  public static final String TAG = "AndroidBrowserHistoryDataExtender";
-  
+  public static final String LOG_TAG = "A.BrowHist.DataExtender";
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/sync/repositories/android/AndroidBrowserHistoryDataExtender.java]: Let's go with <tt>SyncHistoryVisits</tt>.

File: src/main/java/org/mozilla/gecko/sync/repositories/android/AndroidBrowserHistoryDataExtender.java [hunk 2]
```diff
@@ -123,15 +126,29 @@ public void wipe() {
...
+   * Store visit data.
+   *
+   * If a row with GUID `guid` does not exist, insert a new row.
+   * If a row with GUID `guid` does exist, replace the visits column.
+   *
+   * @param guid The GUID to store to.
+   * @param visits New visits data.
+   */
+  public void store(String guid, JSONArray visits) {
     SQLiteDatabase db = this.getCachedReadableDatabase();
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/sync/repositories/android/AndroidBrowserHistoryDataExtender.java]: _blink_

I'm pretty sure that should be a WritableDatabase…

File: src/main/java/org/mozilla/gecko/sync/repositories/android/AndroidBrowserHistoryDataExtender.java [hunk 3]
```diff
@@ -139,34 +156,80 @@ public long store(String guid, JSONArray visits) {
     } else {
       cv.put(COL_VISITS, visits.toJSONString());
     }
-    long rowId = db.insert(TBL_HISTORY_EXT, null, cv);
-    Log.i(TAG, "Inserted history extension record into row: " + rowId);
-    return rowId;
+
+    if (!rowExists) {
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/sync/repositories/android/AndroidBrowserHistoryDataExtender.java]: No need for this. Do something like this:

<pre>
int updated = db.update(TBL_HISTORY_EXT, cv, where, args);
if (updated > 0) {
  // Log
  return;
}
db.insert(TBL_HISTORY_EXT, null, cv);
</pre>


The update _is_ a query, followed by an in-place update on the cursor within the same transaction. Might as well save the round-trip and such.

File: src/main/java/org/mozilla/gecko/sync/repositories/android/AndroidBrowserHistoryDataExtender.java [hunk 4]
```diff
@@ -139,34 +156,80 @@ public long store(String guid, JSONArray visits) {
...
-                          new String[] { COL_GUID, COL_VISITS },
-                          where, args,
-                          null, null, null);
-    RepoUtils.queryTimeLogger("AndroidBrowserHistoryDataExtender.fetch(guid)", queryStart, System.currentTimeMillis());
-    if (cur == null) {
-      Log.e(TAG, "Got a null cursor while doing fetch for guid " + guid + " on history extension table");
-      throw new NullCursorException(null);
-    }
+    Cursor cur = queryHelper.safeQuery(db, ".fetch", TBL_HISTORY_EXT,
+        new String[] { COL_GUID, COL_VISITS },
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/sync/repositories/android/AndroidBrowserHistoryDataExtender.java]: Break this out into a const. No need to allocate a new array.

File: src/main/java/org/mozilla/gecko/sync/repositories/android/AndroidBrowserHistoryDataExtender.java [hunk 5]
```diff
@@ -139,34 +156,80 @@ public long store(String guid, JSONArray visits) {
...
+  /**
+   * Fetch all rows.
+   *
+   * @return A Cursor.
+   * @throws NullCursorException
+   */
+  public Cursor fetchAll() throws NullCursorException {
+    SQLiteDatabase db = this.getCachedReadableDatabase();
+    Cursor cur = db.query(TBL_HISTORY_EXT,
+        new String[] { COL_GUID, COL_VISITS },
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/sync/repositories/android/AndroidBrowserHistoryDataExtender.java]: Again, const.

File: src/main/java/org/mozilla/gecko/sync/repositories/android/AndroidBrowserHistoryDataExtender.java [hunk 6]
```diff
@@ -139,34 +156,80 @@ public long store(String guid, JSONArray visits) {
...
+        null, null, null, null, null);
+    if (cur == null) {
+      throw new NullCursorException(null);
```
Comment (rnewman) [inline src/main/java/org/mozilla/gecko/sync/repositories/android/AndroidBrowserHistoryDataExtender.java]: Can we just use <tt>safeQuery</tt> here?

File: test/src/org/mozilla/gecko/sync/repositories/android/test/AndroidBrowserHistoryDataExtenderTest.java [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+
+import android.content.Context;
+import android.database.Cursor;
+import android.test.ActivityInstrumentationTestCase2;
+import android.util.Log;
+
+public class AndroidBrowserHistoryDataExtenderTest extends ActivityInstrumentationTestCase2<StubActivity> {
+
+  protected AndroidBrowserHistoryDataExtender extender;
+  protected static final String LOG_TAG = "AndroidBrowserHistoryDataExtenderTest";
```
Comment (rnewman) [inline test/src/org/mozilla/gecko/sync/repositories/android/test/AndroidBrowserHistoryDataExtenderTest.java]: Needs to be shorter than 23 characters.

**Pull request merged**